lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -818,6 +828,10 @@ matchWitness(TypeChecker &tc,
     // Check that the mutating bit is ok.
     if (checkMutating(funcReq, funcWitness, funcWitness))
       return RequirementMatch(witness, MatchKind::MutatingConflict);
+    if (funcWitness->isNonMutating() && funcReq->isConsuming())
```

Comment:
SILGen should be able to emit the correct witness thunk for this case -- it would just have to destroy 'self' after calling the witness